src/il_representations/algos/decoders.py

```diff
@@ -276,10 +274,12 @@ def forward(self, z, traj_info, extra_context=None):
         # For Jigsaw, the z input dimension depends on state observation shape;
```

Comment:
This is because Jigsaw has a different way to pass an image. Its forward pass is like this (assuming cutting an image into 3x3 tiles):

```
tiles = image_to_tiles(images)  # [B, C, H, W] -> [B, 9, C, H/3, W/3]

encoder_output = []
for i in range(9):
    tile = x[:, i, :, :, :]
    encoder_representation = self.forward(tile)   # [B, dim]
    encoder_output.append(encoder_representation)
encoder_output = torch.cat(encoder_output, axis=1)  # [B, dim*9]
...
decoder_output = self.decoder(encoder_output)
```

The `encoder_output` is obtained by concatenating 9 `encoder_representation` generated by the encoder network. This is different from other algorithms, where we obtain `encoder_output` just from one network forward pass. This can induce several changes:

1. With the same convolution layers, image tiles with [C, H/3, W/3] produce a different output dimension from an image with shape [C, H, W]. This means if we add a linear layer after conv layers in the encoder during RepL, this linear layer can't be reused for IL. Example:
```
# In encoder.py
t = torch.rand([64, 9, 31, 31]).cuda()
self.forward(t, traj_info=None)
>>> Independent(Normal(loc: torch.Size([64, 128]), scale: torch.Size([64, 128])), 1)

t = torch.rand([64, 9, 100, 100]).cuda()   # DMC tasks
self.forward(t, traj_info=None)
>>>Independent(Normal(loc: torch.Size([64, 1568]), scale: torch.Size([64, 1568])), 1)
```
So instead of having a linear layer in the encoder and discard it during IL, I just directly set the encoder to contain conv layers only, and let the decoder handle the linearity.

2. Since the `encoder_output` shape eventually depends on `observation_space` and `encoder.network`'s structure instead of a preset `representation_dim`, we either set the decoder `projection_layers` by passing in `observation_space` and `encoder.network` as a kwarg during initialization or adjust it upon receiving an input. I think adjusting it on the fly will make fewer changes to the code, so I wrote the code block starting L274 above. To avoid constantly changing `self.projection_layers`, I just added an assertion so it is only allowed to change once (i.e. when it first receives an input).

Reply:
Thanks for the thorough explanation! I looked at fig 3 of the paper & it seems that this is how they do it too. The main thing I'm concerned about is someone saving the model, reloading it, evaluating on a different dataset, and having it silently reset its internal weights. Now that we have the assertion I think we should be pretty good; I might add a warning too so that the user can see what is going on (& decide whether it is desirable).